common-app-lib/src/cmr/common_app/app_events.clj

```diff
@@ -0,0 +1,74 @@
+(ns cmr.common-app.app-events
+  "Common events which applications can use when starting and stopping."
+  (:require
+   [clojure.java.io :as io]
+   [cmr.common-app.config :as config]
+   [cmr.common.log :refer [debugf infof warnf errorf]]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(import 'com.sun.management.HotSpotDiagnosticMXBean)
+(import 'java.lang.management.ManagementFactory)
+
+;;
+;; The internet suggests the following as how one would write the dump file in java:
+;;
+;;public static void dumpHeap (String filePath, boolean live) throws IOException {
+;;  MBeanServer server = ManagementFactory.getPlatformMBeanServer ();
+;;  HotSpotDiagnosticMXBean mxBean = ManagementFactory.newPlatformMXBeanProxy (server,
+;;      "com.sun.management:type=HotSpotDiagnostic", HotSpotDiagnosticMXBean.class);
+;;   mxBean.dumpHeap (filePath, live);
+;;}
+
+(defn- dump-heap
+  "Write a diagnostics file to the file system for analysis in Visual VM"
+  [^String app-name ^String file-path ^java.lang.Boolean live]
+  (let [bean (-> (ManagementFactory/getPlatformMBeanServer)
+                 (ManagementFactory/newPlatformMXBeanProxy "com.sun.management:type=HotSpotDiagnostic"
+                                                           HotSpotDiagnosticMXBean))]
+    (try
+      (when (.exists (io/file file-path))
+        (warnf (format "dump-heap: %s deleting diagnostic file %s." app-name file-path))
+        (io/delete-file file-path))
+      (.dumpHeap bean file-path live)
+      (infof "dump-heap: %s finished creating diagnostic file %s." app-name file-path)
+      (catch Exception err
+        (errorf "dump-heap: %s because %s" app-name (.getMessage err))))))
+
+(defn- writeDump
+  "A call back function which writes a diagnostics file if the application config
+   dump-diagnostics-on-exist-to has been set to a file path. To test this out,
+   run the following commands:
+   >CMR_DUMP_DIAGNOSTICS_ON_EXIT_TO='/tmp/dump-test.hprof'
+       java -cp cmr-search-app-0.1.0-SNAPSHOT-standalone.jar clojure.main
+       -m cmr.search.runner
+   >kill -15 $(jps | grep main | cut -f 1 -d ' ')"
+  [app-name]
+  (let [path-to-file (config/dump-diagnostics-on-exit-to)]
+    (infof "dump-heap: %s is writing a diagnostic file to %s..." app-name path-to-file)
+    (dump-heap app-name path-to-file false)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn stop-on-exit-hook
```

Comment:
Will using this hook function make sure that the stop command is actually run in the microservices? I know that until now searching for "system stopped" in logs doesn't really show anything

Reply:
yes, it should, unless something uses either kill -9 or the Task is blown away at the top. AWS should try to shutdown nicely, if that is the case we will see our shutdown messages. Locally I test this with `kill -15`.

Reply:
We will know when we deploy and then re-deploy, initiating shutdown. We may not know what happens in "bad" situations until one occurs naturally however.

Reply:
right. I'm very interested in seeing how this behaves in the real world.